x/protorev/keeper/hooks_test.go

```diff
@@ -0,0 +1,315 @@
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.ExitSwapExactAmountOut(s.Ctx, s.TestAccs[0], 1, sdk.NewCoin("Atom", sdk.NewInt(1)), sdk.NewInt(1002141106353159235))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "GAMM - Exit Swap Share Amount In",
+			param: param{
+				expectedTrades: []types.Trade{
+					{
+						Pool:     1,
+						TokenIn:  "akash",
+						TokenOut: "Atom",
+					},
+				},
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.ExitSwapShareAmountIn(s.Ctx, s.TestAccs[0], 1, "Atom", sdk.NewInt(1000000000000000000), sdk.NewInt(1))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "GAMM - Exit Swap Share Amount In - Low Shares",
+			param: param{
+				expectedTrades: []types.Trade(nil),
+				executeLiquidityProviding: func() {
+					_, err := s.App.GAMMKeeper.ExitSwapShareAmountIn(s.Ctx, s.TestAccs[0], 1, "Atom", sdk.NewInt(1000), sdk.NewInt(0))
+					s.Require().NoError(err)
+				},
+			},
+			expectPass: true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			tc.param.executeLiquidityProviding()
+
+			routes, err := s.App.ProtoRevKeeper.GetSwapsToBackrun(s.Ctx)
+			s.Require().NoError(err)
+			s.Require().Equal(tc.param.expectedTrades, routes.Trades)
+
+			s.App.ProtoRevKeeper.DeleteSwapsToBackrun(s.Ctx)
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestPoolCreation() {
+	type param struct {
+		matchDenom          string
+		executePoolCreation func() uint64
+	}
+
+	tests := []struct {
+		name       string
+		param      param
+		expectPass bool
+	}{
+		{
+			name: "GAMM - Create Pool",
+			param: param{
+				matchDenom: "hookGamm",
+				executePoolCreation: func() uint64 {
+					poolId := s.createGAMMPool([]balancer.PoolAsset{
+						{
+							Token:  sdk.NewCoin("hookGamm", sdk.NewInt(1000000000)),
+							Weight: sdk.NewInt(1),
+						},
+						{
+							Token:  sdk.NewCoin(types.OsmosisDenomination, sdk.NewInt(1000000000)),
+							Weight: sdk.NewInt(1),
+						},
+					},
+						sdk.NewDecWithPrec(2, 3),
+						sdk.NewDecWithPrec(0, 2))
+
+					return poolId
+				},
+			},
+			expectPass: true,
+		},
+		{
+			name: "Concentrated Liquidity - Create Pool w/ No Liqudity",
+			param: param{
+				matchDenom: "hookCL",
+				executePoolCreation: func() uint64 {
+					clPool := s.PrepareConcentratedPool()
+					return clPool.GetId()
+				},
+			},
+			expectPass: false,
+		},
+		{
+			name: "Concentrated Liquidity - Create Pool w/ Liqudity",
+			param: param{
+				matchDenom: "hookCL",
+				executePoolCreation: func() uint64 {
+					clPool := s.PrepareConcentratedPoolWithCoinsAndFullRangePosition("hookCL", "uosmo")
+					return clPool.GetId()
+				},
+			},
+			expectPass: true,
+		},
+	}
+
+	for _, tc := range tests {
+		s.Run(tc.name, func() {
+			poolId := tc.param.executePoolCreation()
```

Comment:
I would like to see a test where CL and gamm pools created have the same denoms.

1. Balancer is created first and CL second
2. CL is created first and balancer second

Reply:
Makes sense, added here: https://github.com/osmosis-labs/osmosis/pull/5045/commits/c24c3b98f9c987f36a539713bc196d09cc829b52